src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs

```diff
@@ -10,4 +10,402 @@
+      pushStack (VStack.LamPartialE lamPartial)
+      pure (Env.Curr lamPartial)
+    Just (VStack.Position usage index)
+      | one == usage →
+        Env.Expanded <$> moveToFront index
+      | otherwise →
+        Env.Expanded <$> dupToFront index
+
+-- |
+-- Name calls inst, and then determines how best to name the form in the VStack
+name ∷ Env.Reduction m ⇒ Symbol → Types.NewTerm → m Env.Expanded
+name symb f@(form, _usage, _type') = do
+  result ← inst f
+  case form of
+    Ann.Var {} →
+      modify @"stack" (VStack.nameTop symb)
+    -- all prims shouldn't add their arguments to the vstack
+    Ann.Prim {} →
+      modify @"stack" (VStack.nameTop symb)
+  -- consVar symb result usage type'
+  pure result
+
+-- for constant we shouldn't be applying it unless it's a lambda Ι don't think!?
+primToFargs ∷ Num b ⇒ Types.NewPrim → Types.Type → (Env.Fun, b)
+primToFargs (Types.Constant _) _ = undefined
+primToFargs (Types.Inst inst) ty =
+  case inst of
+    -- Can't abstract out pattern due to bad forall resolution!
+    Instr.ADD _ → (Env.Fun (add ty), 2)
+    Instr.SUB _ → (Env.Fun (sub ty), 2)
+    Instr.MUL _ → (Env.Fun (mul ty), 2)
+    Instr.EDIV _ → (Env.Fun (ediv ty), 2)
+    Instr.OR {} → (Env.Fun (or ty), 2)
+    Instr.AND _ → (Env.Fun (and ty), 2)
+    Instr.XOR _ → (Env.Fun (xor ty), 2)
+
+appM ∷ Env.Reduction m ⇒ Types.NewTerm → [Types.NewTerm] → m Env.Expanded
+appM form@(t, _u, ty) args =
+  case t of
+    Ann.Prim p →
+      let (f, lPrim) = primToFargs p ty
+          argsL = length args
+       in case argsL `compare` lPrim of
+            EQ → Env.unFun f args
+            LT → do
+              names ← reserveNames (fromIntegral lPrim)
+              -- TODO ∷ set the usage of the arguments to 1
+              let constructed = Env.C
+                    { Env.fun = f,
+                      Env.argsLeft = names,
+                      Env.left = fromIntegral lPrim,
+                      Env.captures = Set.empty,
+                      Env.ty = ty,
+                      Env.argsApplied = []
+                    }
+              apply constructed args []
+            -- this should never happen, due to type checking??
+            GT →
+              throw
+                @"compilationError"
+                (Types.InternalFault "Michelson call with too many args")
+    Ann.LamM captures arguments body →
+      let constructed = Env.C
+            { Env.fun = Env.Fun (const (inst body)),
+              Env.argsLeft = arguments,
+              Env.captures = Set.fromList captures,
+              Env.left = fromIntegral (length arguments),
+              Env.ty = ty,
+              Env.argsApplied = []
+            }
+       in apply constructed args []
+    Ann.Var _ → do
+      v ← inst form
+      case v of
+        Env.Curr c → apply c args []
+        -- these two cases would only be valid if we expanded into a Michelson
+        -- lambda, however would we ever do that?
```

Comment:
As written, no, however we might have *variables* which are Michelson lambdas (e.g. passed as function arguments), do we need to handle that case here? Seems like we might.

Reply:
We do handle this, if we are applying functions, we do not want a Michelson lambda. When a lambda is an argument to a primtive we promote it. The only other time a Michelson Lambda gets made is if we don't have enough arguments at the end of all these applications

Reply:
Note I don't have the logic for promote for lambdas online, I was hoping to delegate this to you, which would be using the DSL to cleanup and replace the existing one!

Reply:
Hmm, I don't quite follow - if lambdas are passed in as arguments, they will be equivalent (in core / the higher-level language) to native functions, so we will be *applying* them to things - do you mean that we'll promote them to this `Env.Curr` form? I'm not sure that we can, and in any case we will need logic to actually call `EXEC`, which we wouldn't do with `Env.Curr`, right?

Reply:
Well if we pass them as an argument, to say map, then they should be `promoted` from `Env.Curr` to a real Michelson Lambda.

However if we are just passing around a function in the sense of
```ocaml
let f g = g 3
```
then, we would compile g as a `Env.Curr`, and then in f we will look up g (which is an `Env.Curr`) and then go through the apply logic.

if this is the top level and g takes two arguments, then we would `promote` this into a Michelson lambda.

As per calling `EXEC`, we have all the instructions in the Environment, so you could temporarly empty it, run the capture logic you need to, then create a Michelson lambda.


`Env.Curr` Does not deal with this as this is not `Env.Curr's` responsability

Reply:
Right, I am wondering what we would do in the case of a lambda passed as an argument to a function in our contract (so one that we do not compile, but that we do evaluate).

Reply:
It would compile correctly, as it'd dispatch to the application case on var, which looks it up from the env, and sees if we are LT GT or EQ in respect to how many arguments are left.

Thus if it has EQ number of args, it'd inline completely. There may be some slowdown if it's 

let f a = (+) a, since we have to move names a bit too much (I have this exact case written on paper) in the shuffles, but otherwise be optimal

Reply:
Right, but in that case when do we use `EXEC`? `EXEC` isn't required for (normal) inlined functions which we're compiling (and aren't being provided as arguments), but it *is* required for lambda functions passed as arguments.

Reply:
Ahh that's what you meant. This happens when we promote the top of the VStack. when we see a function like map, it'll put the top two arguments on the stack and promote them, meaning our lambda will turn into a real compiled lambda.

This is the logic I hope you to write (as it's the promotoing our curried -> Michelson Lambda)

Reply:
This is for the case of external parameters, not functions we compile - we should change [this](https://github.com/cryptiumlabs/juvix/blob/07ec900f4ece3e57d8b4eed5de66fcf8a3eb91c2/src/Juvix/Backends/Michelson/DSL/InstructionsEff.hs#L178).

Reply:
We can expect external Michelson functions to either be curried or use an explicit tuple type.